dependency-specification.rst

```diff
@@ -0,0 +1,543 @@
+    version       = wsp* ( letterOrDigit | '-' | '_' | '.' | '*' )+
+    version_one   = version_cmp version wsp*
+    version_many  = version_one (wsp* ',' version_one)*
+    versionspec   = ( '(' version_many ')' ) | version_many
+    urlspec       = '@' wsp* <URI_reference>
+
+Environment markers allow making a specification only take effect in some
+environments::
+
+    marker_op     = version_cmp | 'in' | 'not' wsp+ 'in'
+    python_str_c  = (wsp | letter | digit | '(' | ')' | '.' | '{' | '}' |
+                     '-' | '_' | '*')
+    dquote        = '"'
+    squote        = '\\''
+    python_str    = (squote (python_str_c | dquote)* squote |
+                     dquote (python_str_c | squote)* dquote)
+    env_var       = ('python_version' | 'python_full_version' |
+                     'os_name' | 'sys_platform' | 'platform_release' |
+                     'platform_system' | 'platform_version' |
+                     'platform_machine' | 'python_implementation' |
+                     'implementation_name' | 'implementation_version' |
+                     'extra' # ONLY when defined by a containing layer
+                     )
+    marker_var    = env_var | python_str
+    marker_expr   = ('(' wsp* marker wsp* ')'
+                     | (marker_var wsp* marker_op wsp* marker_var))
+    marker        = wsp* marker_expr ( wsp* ('and' | 'or') wsp* marker_expr)*
+    quoted_marker = ';' wsp* marker
+
+Optional components of a distribution may be specified using the extras
+field::
+
+    identifier    = letterOrDigit (
+                    letterOrDigit |
+                    (( letterOrDigit | '-' | '_' | '.')* letterOrDigit ) )*
+    name          = identifier
+    extras_list   = identifier (wsp* ',' wsp* identifier)*
+    extras        = '[' wsp* extras_list? wsp* ']'
+
+Giving us a rule for name based requirements::
+
+    name_req      = name wsp* extras? wsp* versionspec? wsp* quoted_marker?
+
+And a rule for direct reference specifications::
+
+    url_req       = name wsp* extras? wsp* urlspec wsp+ quoted_marker?
+
+Leading to the unified rule that can specify a dependency.::
+
+    specification = wsp* ( url_req | name_req ) wsp*
+
+Whitespace
+----------
+
+Non line-breaking whitespace is mostly optional with no semantic meaning. The
+sole exception is detecting the end of a URL requirement.
+
+Names
+-----
+
+Python distribution names are currently defined in PEP-345 [#pep345]_. Names
+act as the primary identifier for distributions. They are present in all
+dependency specifications, and are sufficient to be a specification on their
+own. However, PyPI places strict restrictions on names - they must match a
+case insensitive regex or they won't be accepted. Accordingly in this PEP we
+limit the acceptable values for identifiers to that regex. A full redefinition
+of name may take place in a future metadata PEP::
+
+    ^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$
```

Comment:
I want to use the exact regex PyPI uses; should I add some prose or a human description?